flink-end-to-end-tests/flink-datastream-allround-test/src/main/java/org/apache/flink/streaming/tests/SlidingWindowCheckMapper.java

```diff
@@ -0,0 +1,178 @@
+		List<Tuple2<Event, Integer>> previousWindowValues = Optional.ofNullable(eventsSeenSoFar.value()).orElseGet(
+			Collections::emptyList);
+
+		List<Event> newValues = value.f1;
+		Optional<Event> lastEventInWindow = verifyWindowContiguity(newValues, out);
+
+		Long lastSequenceNumberSeenSoFar = lastSequenceNumber.value();
+		List<Tuple2<Event, Integer>> newWindows =
+			verifyPreviousOccurences(previousWindowValues, newValues, out, lastSequenceNumberSeenSoFar);
+
+		if (lastEventInWindow.isPresent()) {
+			updateLastSeenSequenceNumber(lastEventInWindow.get(), lastSequenceNumberSeenSoFar);
+		}
+
+		eventsSeenSoFar.update(newWindows);
+	}
+
+	private void updateLastSeenSequenceNumber(
+			Event lastEventInWindow,
+			Long lastSequenceNumberSeenSoFar) throws IOException {
+		long lastSequenceNumberInWindow = lastEventInWindow.getSequenceNumber();
+		if (lastSequenceNumberSeenSoFar == null || lastSequenceNumberInWindow > lastSequenceNumberSeenSoFar) {
+			lastSequenceNumber.update(lastSequenceNumberInWindow);
+		}
+	}
+
+	/**
+	 * Verifies if all values from previous windows appear in the new one. Returns union of all events seen so far that
+	 * were not seen <b>slideFactor</b> number of times yet.
+	 */
+	private List<Tuple2<Event, Integer>> verifyPreviousOccurences(
+			List<Tuple2<Event, Integer>> previousWindowValues,
+			List<Event> newValues,
+			Collector<String> out, Long lastSequenceNumberSeenSoFar) {
+		List<Tuple2<Event, Integer>> newEventsSeenSoFar = new ArrayList<>();
+		List<Event> seenWindows = new ArrayList<>();
+
+		for (Tuple2<Event, Integer> windowValue : previousWindowValues) {
+			if (!newValues.contains(windowValue.f0)) {
+				printEventNotSeenAlertMessage(windowValue, newValues, out);
+			} else {
+				seenWindows.add(windowValue.f0);
+				preserveOrDiscardIfSeenSlideFactorTimes(newEventsSeenSoFar, windowValue);
+			}
+		}
+
+		addNotSeenValues(newEventsSeenSoFar, newValues, seenWindows, lastSequenceNumberSeenSoFar, out);
+
+		return newEventsSeenSoFar;
+	}
+
+	private void addNotSeenValues(
+			List<Tuple2<Event, Integer>> newEventsSeenSoFar,
+			List<Event> newValues,
+			List<Event> seenWindows,
+			Long lastSequenceNumberSeenSoFar,
+			Collector<String> out) {
+		newValues.stream()
+			.filter(e -> !seenWindows.contains(e))
+			.forEach(e -> {
+				if (lastSequenceNumberSeenSoFar == null || e.getSequenceNumber() > lastSequenceNumberSeenSoFar) {
+					newEventsSeenSoFar.add(Tuple2.of(e, 1));
+				} else {
+					printEventSeenTooManyTimesMessage(e, out);
+				}
+			});
+	}
+
+	private void printEventSeenTooManyTimesMessage(Event e, Collector<String> out) {
+		out.collect(String.format("Alert: event %s seen more than %d times", e, slideFactor));
+	}
+
+	private void preserveOrDiscardIfSeenSlideFactorTimes(
+			List<Tuple2<Event, Integer>> newEvenstSeenSoFar,
+			Tuple2<Event, Integer> windowValue) {
+		if (windowValue.f1 + 1 != slideFactor) {
```

Comment:
For clarity I would put: `int timesSeen = windowValue.f1;` and then use the `timesSeen` in the next line.